// @react-three/drei/core/Outlines.js
import _extends from '@babel/runtime/helpers/esm/extends';
import * as THREE from 'three';
import * as React from 'react';
import { shaderMaterial } from './shaderMaterial.js';
import { extend, applyProps } from '@react-three/fiber';
import { toCreasedNormals } from 'three-stdlib';

const OutlinesMaterial = shaderMaterial({
  color: new THREE.Color('black'),
  opacity: 1,
  thickness: 0.05
}, `#include <common>
   #include <morphtarget_pars_vertex>
   #include <skinning_pars_vertex>
   uniform float thickness;
   void main() {
     #if defined (USE_SKINNING)
	     #include <beginnormal_vertex>
       #include <morphnormal_vertex>
       #include <skinbase_vertex>
       #include <skinnormal_vertex>
       #include <defaultnormal_vertex>
     #endif
     #include <begin_vertex>
	   #include <morphtarget_vertex>
	   #include <skinning_vertex>
     #include <project_vertex>
     vec4 transformedNormal = vec4(normal, 0.0);
     vec4 transformedPosition = vec4(transformed, 1.0);
     #ifdef USE_INSTANCING
       transformedNormal = instanceMatrix * transformedNormal;
       transformedPosition = instanceMatrix * transformedPosition;
     #endif
     vec3 newPosition = transformedPosition.xyz + transformedNormal.xyz * thickness;
     gl_Position = projectionMatrix * modelViewMatrix * vec4(newPosition, 1.0); 
   }`, `uniform vec3 color;
   uniform float opacity;
   void main(){
     gl_FragColor = vec4(color, opacity);
     #include <tonemapping_fragment>
     #include <${parseInt(THREE.REVISION.replace(/\D+/g, '')) >= 154 ? 'colorspace_fragment' : 'encodings_fragment'}>
   }`);
function Outlines({
  color = 'black',
  opacity = 1,
  transparent = false,
  thickness = 0.05,
  angle = Math.PI,
  ...props
}) {
  var _ref$current, _ref$current$parent;

  const ref = React.useRef(null);
  const [material] = React.useState(() => new OutlinesMaterial({
    side: THREE.BackSide
  }));
  React.useMemo(() => extend({
    OutlinesMaterial
  }), []);
  React.useLayoutEffect(() => {
    const group = ref.current;
    const parent = group.parent;

    if (parent && parent.geometry) {
      let mesh;

      if (parent.skeleton) {
        mesh = new THREE.SkinnedMesh();
        mesh.material = material;
        mesh.bind(parent.skeleton, parent.bindMatrix);
        group.add(mesh);
      } else if (parent.isInstancedMesh) {
        mesh = new THREE.InstancedMesh(parent.geometry, material, parent.count);
        mesh.instanceMatrix = parent.instanceMatrix;
        group.add(mesh);
      } else {
        mesh = new THREE.Mesh();
        mesh.material = material;
        group.add(mesh);
      }

      mesh.geometry = angle ? toCreasedNormals(parent.geometry, angle) : parent.geometry;
      return () => {
        if (angle) mesh.geometry.dispose();
        group.remove(mesh);
      };
    }
  }, [angle, (_ref$current = ref.current) == null ? void 0 : (_ref$current$parent = _ref$current.parent) == null ? void 0 : _ref$current$parent.geometry]);
  React.useLayoutEffect(() => {
    const group = ref.current;
    const mesh = group.children[0];

    if (mesh) {
      applyProps(mesh.material, {
        transparent,
        thickness,
        color,
        opacity
      });
    }
  }, [angle, transparent, thickness, color, opacity]);
  return /*#__PURE__*/React.createElement("group", _extends({
    ref: ref
  }, props));
}

export { Outlines };
